elements/lisk-tree/src/merkle_tree/merkle_tree.ts

```diff
@@ -267,13 +276,100 @@ export class MerkleTree {
 		return rightWitness;
 	}
 
+	public async update(idxs: number[], updateData: ReadonlyArray<Buffer>): Promise<Buffer> {
+		const updateHashes = [];
+		for (const data of updateData) {
+			const leafValueWithoutNodeIndex = Buffer.concat(
+				[LEAF_PREFIX, data],
+				LEAF_PREFIX.length + data.length,
+			);
+			const leafHash = hash(leafValueWithoutNodeIndex);
+			updateHashes.push(leafHash);
+		}
+		const siblingHashes = await this._getSiblingHashes(idxs);
+
+		const calculatedTree = calculatePathNodes(updateHashes, this._size, idxs, siblingHashes);
+		const height = this._getHeight();
+
+		for (const [index, hashedValue] of calculatedTree.entries()) {
+			const loc = getLocation(index, height);
+			const nodeIndexBuffer = Buffer.alloc(NODE_INDEX_SIZE);
+			nodeIndexBuffer.writeInt32BE(loc.nodeIndex, 0);
+			const prefix = loc.layerIndex === 0 ? LEAF_PREFIX : BRANCH_PREFIX;
+			const value = Buffer.concat(
+				[prefix, nodeIndexBuffer, hashedValue],
+				prefix.length + nodeIndexBuffer.length + hashedValue.length,
+			);
+			await this._hashToValueMap.set(hashedValue, value);
+			await this._locationToHashMap.set(getBinaryString(loc.nodeIndex, height), hashedValue);
+		}
+
+		return calculatedTree.get(ROOT_INDEX) as Buffer;
+	}
+
 	public async toString(): Promise<string> {
 		if (this._size === 0) {
 			return this.root.toString('hex');
 		}
 		return this._printNode(this.root);
 	}
 
+	private async _getSiblingHashes(idxs: number[]): Promise<Buffer[]> {
+		let sortedIdxs: number[] = [...idxs];
+		sortedIdxs.sort(treeSortFn);
+		const siblingHashes: Buffer[] = [];
+		const size = this._size;
+		const height = this._getHeight();
+		while (sortedIdxs.length > 0) {
+			const currentIndex = sortedIdxs[0];
+			// check for next index in case I can use it if: node is left AND there are other indices AND next index is at distance 1 AND it is on the same layer
+			// in that case remove it from the indices
+			if (
+				isLeft(currentIndex) &&
+				sortedIdxs.length > 1 &&
+				isSameLayer(currentIndex, sortedIdxs[1]) &&
+				areSiblings(currentIndex, sortedIdxs[1])
+			) {
+				sortedIdxs = sortedIdxs.slice(2);
+				const parentIndex = currentIndex >> 1;
+				insertNewIndex(sortedIdxs, parentIndex);
+				continue;
+			}
+			const currentNodeLoc = getLocation(currentIndex, height);
+			if (currentNodeLoc.layerIndex === height) {
+				return siblingHashes;
+			}
+			const siblingNode = getRightSiblingInfo(
```

Comment:
Should we consider renaming the function `getRightSiblingInfo`?

Reply:
any suggestion? I think this gets the sibling information except from left up right? Ill add the comment as well